pkg/resources/openstack/resources.go

```diff
@@ -17,16 +17,220 @@ limitations under the License.
 package openstack
 
 import (
+	"fmt"
+	"strings"
+
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/routers"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/subnets"
+
+	"github.com/golang/glog"
 	"k8s.io/kops/pkg/resources"
+	"k8s.io/kops/upup/pkg/fi"
 	"k8s.io/kops/upup/pkg/fi/cloudup/openstack"
 )
 
-type listFn func(openstack.OpenstackCloud, string) ([]*resources.Resource, error)
+type openstackListFn func() ([]*resources.Resource, error)
+
+const (
+	typeSSHKey   = "SSHKey"
+	typeRouterIF = "Router-IF"
+	typeRouter   = "Router"
+	typeSubnet   = "Subnet"
+	typeNetwork  = "Network"
+)
 
 func ListResources(cloud openstack.OpenstackCloud, clusterName string) (map[string]*resources.Resource, error) {
-	resourceTrackers := make(map[string]*resources.Resource)
+	resources := make(map[string]*resources.Resource)
+
+	os := &clusterDiscoveryOS{
+		cloud:       cloud,
+		osCloud:     cloud,
+		clusterName: clusterName,
+	}
+
+	listFunctions := []openstackListFn{
+		os.ListKeypairs,
+		os.ListNetwork,
+	}
+	for _, fn := range listFunctions {
+		resourceTrackers, err := fn()
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range resourceTrackers {
+			resources[t.Type+":"+t.ID] = t
+		}
+	}
+	return resources, nil
+}
+
+type clusterDiscoveryOS struct {
+	cloud       fi.Cloud
+	osCloud     openstack.OpenstackCloud
+	clusterName string
+
+	zones []string
+}
+
+func openstackKeyPairName(org string) string {
+	name := strings.Replace(org, ".", "-", -1)
+	name = strings.Replace(name, ":", "_", -1)
+	return name
+}
+
+func (os *clusterDiscoveryOS) ListNetwork() ([]*resources.Resource, error) {
```

Comment:
Can we move these specific tasks to their own files?  

Working on doing this as well for what has already been implemented in cloud.go
https://github.com/cisco-sso/kops/tree/openstack_model/upup/pkg/fi/cloudup/openstack

Reply:
yes we can